source/interface/azure_iot_mqtt.h

```diff
@@ -0,0 +1,380 @@
+#define azureiotmqttPACKET_TYPE_UNSUBSCRIBE    ( ( uint8_t ) 0xA2U )  /** @brief UNSUBSCRIBE (client-to-server). */
+#define azureiotmqttPACKET_TYPE_UNSUBACK       ( ( uint8_t ) 0xB0U )  /** @brief UNSUBACK (server-to-client). */
+#define azureiotmqttPACKET_TYPE_PINGREQ        ( ( uint8_t ) 0xC0U )  /** @brief PINGREQ (client-to-server). */
+#define azureiotmqttPACKET_TYPE_PINGRESP       ( ( uint8_t ) 0xD0U )  /** @brief PINGRESP (server-to-client). */
+#define azureiotmqttPACKET_TYPE_DISCONNECT     ( ( uint8_t ) 0xE0U )  /** @brief DISCONNECT (client-to-server). */
+
+
+typedef enum AzureIoTMQTTQoS
+{
+    eAzureIoTMQTTQoS0 = 0, /** Delivery at most once. */
+    eAzureIoTMQTTQoS1 = 1, /** Delivery at least once. */
+    eAzureIoTMQTTQoS2 = 2  /** Delivery exactly once. */
+} AzureIoTMQTTQoS_t;
+
+typedef enum AzureIoTMQTTSubAckStatus
+{
+    eMQTTSubAckSuccessQos0 = 0x00, /** Success with a maximum delivery at QoS 0. */
+    eMQTTSubAckSuccessQos1 = 0x01, /** Success with a maximum delivery at QoS 1. */
+    eMQTTSubAckSuccessQos2 = 0x02, /** Success with a maximum delivery at QoS 2. */
+    eMQTTSubAckFailure = 0x80      /** Failure. */
+} AzureIoTMQTTSubAckStatus_t;
+
+typedef enum AzureIoTMQTTResult
+{
+    eAzureIoTMQTTSuccess = 0,     /** Function completed successfully. */
+    eAzureIoTMQTTBadParameter,    /** At least one parameter was invalid. */
+    eAzureIoTMQTTNoMemory,        /** A provided buffer was too small. */
+    eAzureIoTMQTTSendFailed,      /** The transport send function failed. */
+    eAzureIoTMQTTRecvFailed,      /** The transport receive function failed. */
+    eAzureIoTMQTTBadResponse,     /** An invalid packet was received from the server. */
+    eAzureIoTMQTTServerRefused,   /** The server refused a CONNECT or SUBSCRIBE. */
+    eAzureIoTMQTTNoDataAvailable, /** No data available from the transport interface. */
+    eAzureIoTMQTTIllegalState,    /** An illegal state in the state record. */
+    eAzureIoTMQTTStateCollision,  /** A collision with an existing state record entry. */
+    eAzureIoTMQTTKeepAliveTimeout /** Timeout while waiting for PINGRESP. */
+} AzureIoTMQTTResult_t;
+
+typedef struct AzureIoTMQTTConnectInfo
+{
+    /**
+     * @brief Whether to establish a new, clean session or resume a previous session.
+     */
+    bool xCleanSession;
+
+    /**
+     * @brief MQTT keep alive period.
+     */
+    uint16_t usKeepAliveSeconds;
+
+    /**
+     * @brief MQTT client identifier. Must be unique per client.
+     */
+    const char * pcClientIdentifier;
+
+    /**
+     * @brief Length of the client identifier.
+     */
+    uint16_t usClientIdentifierLength;
+
+    /**
+     * @brief MQTT user name. Set to NULL if not used.
+     */
+    const char * pcUserName;
+
+    /**
+     * @brief Length of MQTT user name. Set to 0 if not used.
+     */
+    uint16_t usUserNameLength;
+
+    /**
+     * @brief MQTT password. Set to NULL if not used.
+     */
+    const char * pcPassword;
```

Comment:
I'm assuming that a const char* here and assumed \0 at end of string is OK, instead of byte/length?  If overwhelming # of stacks use that then it's fine.  I vaguely recall a similar discussion on embedded asnd string OK but want to double check.

Reply:
this is buffer with corresponding length, I can change it to uint8_t, but that will require stdint.h as a dependent header file. 

---
In reply to: [598960913](https://github.com/Azure/azure-iot-middleware-freertos/pull/44#discussion_r598960913) [](ancestors = 598960913)

Reply:
updated

---
In reply to: [599158305](https://github.com/Azure/azure-iot-middleware-freertos/pull/44#discussion_r599158305) [](ancestors = 599158305,598960913)

Reply:
I'm not that concerted about stdint.h since I'm almost positive (and [Wikipdeia](https://en.wikipedia.org/wiki/C_data_types#stdint.h) seems to confirm) that it's a core part of C99.  It's other random headers I worry about.  

I think that updating this is correct here since it keeps consistency up until we call off into the 3rd party MQTT stack, but if that proves wrong we haev flexibility at this layer (for now!) to change.  Thanks!